### ISSUE TITLE
Donwload MNIST under Windows behind Proxy

### ISSUE THREAD

**Issue by mhechthz** [ageron/handson-ml#229]
Dear Aurelien,

I just tried to run some of your examples an fail with MNIST, i.e. the line

`mnist = input_data.read_data_sets("/tmp/data/")`

following error is generated

```
---------------------------------------------------------------------------
PermissionDeniedError                     Traceback (most recent call last)
...
d:\Anaconda3w64\lib\site-packages\tensorflow\python\framework\errors_impl.py in raise_exception_on_not_ok_status()
    464           None, None,
    465           compat.as_text(pywrap_tensorflow.TF_Message(status)),
--> 466           pywrap_tensorflow.TF_GetCode(status))
    467   finally:
    468     pywrap_tensorflow.TF_DeleteStatus(status)

PermissionDeniedError: Failed to create a directory: /tmp

```

I work under Windows, which usually works well. The Proxy is set properly.

If I change the line to 
`mnist = input_data.read_data_sets("./datasets/data/")`
I get
```
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
...
d:\Anaconda3w64\lib\urllib\request.py in do_open(self, http_class, req, **http_conn_args)
   1318                           encode_chunked=req.has_header('Transfer-encoding'))
   1319             except OSError as err: # timeout error
-> 1320                 raise URLError(err)
   1321             r = h.getresponse()
   1322         except:

URLError: <urlopen error Tunnel connection failed: 307 Temporary Redirect>
```

What can I do now?

Best regards,

Michael

**Comment by debasish2256** [comment]
Actually when I run it on data camp integrated coding platform it runs fine
with no error

**Comment by mhechthz** [comment]
I tried it at home, where no proxy runs and all works well. So I think it's an issue of the proxy?

Any idea?

**Comment by ageron** [comment]
Hi @debasish2256 ,
Thanks for your feedback.  The first error "`PermissionDeniedError: Failed to create a directory: /tmp`" is simply due to the fact that there is no `/tmp` directory on Windows.  You can either enter the path to a directory for which you have read/write access (as you did), or you can use the following code that creates a temporary directory:

```python
>>> from tempfile import mkdtemp
>>> tmp_dir = mkdtemp()
>>> import tensorflow.examples.tutorials.mnist.input_data as input_data
>>> input_data.read_data_sets(tmp_dir)
Successfully downloaded train-images-idx3-ubyte.gz 9912422 bytes.
Extracting /var/folders/wy/h39t6kb11pnbb0pzhksd_fqh0000gn/T/tmpucqwnd5y/train-images-idx3-ubyte.gz
Successfully downloaded train-labels-idx1-ubyte.gz 28881 bytes.
Extracting /var/folders/wy/h39t6kb11pnbb0pzhksd_fqh0000gn/T/tmpucqwnd5y/train-labels-idx1-ubyte.gz
Successfully downloaded t10k-images-idx3-ubyte.gz 1648877 bytes.
Extracting /var/folders/wy/h39t6kb11pnbb0pzhksd_fqh0000gn/T/tmpucqwnd5y/t10k-images-idx3-ubyte.gz
Successfully downloaded t10k-labels-idx1-ubyte.gz 4542 bytes.
Extracting /var/folders/wy/h39t6kb11pnbb0pzhksd_fqh0000gn/T/tmpucqwnd5y/t10k-labels-idx1-ubyte.gz
```

The second error "`URLError: <urlopen error Tunnel connection failed: 307 Temporary Redirect>`" is more problematic. It is clearly a network error, and I would guess it's a proxy configuration issue. As @mhechthz mentionned (thanks, by the way), everything works fine when there is no proxy.  In short, I don't think this is a problem linked to TensorFlow, the code, or even python.  I think it's a network configuration issue.  Perhaps check out the new code which uses a different method to download the data.  You can also try to download it directly using your browser.

Hope this helps,
Aurélien

**Comment by ArttuHautakoski** [comment]
Removing the first slash in the path fixed the problem for me. Like this:
 `mnist = input_data.read_data_sets("tmp/data/")`

**Comment by debasish2256** [comment]
Hey,

I just signed the petition "High court : JUSTICE FOR SUPTRISHI" and wanted
to see if you could help by adding your name.

Our goal is to reach 835 signatures and we need more support. You can read
more and sign the petition here:

http://chng.it/Bysdtygt69

Thanks!
Debsish

**Comment by ageron** [comment]
If anyone still has problems downloading MNIST, please check out the fix in #301.
